### PR TITLE
add history feature to review previous bot sessions.

### PR DESCRIPTION
**Pull request by LarryLan**
1) update gitignore to ignore logs and history data
2) add history feature

to do -
1) create a aws s3 handler and upload all the grinding session data to s3
2) create details view for sessions to show items/gold/death etc.

### CODE REVIEW

**Comment by LarryLan** [comment]
dont merge it yet there's a bug when the bot exited unexpectedly for example stucked. the data recorded is wrong. I accidently found it when my character stuck and stopped the bot.

File: BlazorServer/Startup.cs [hunk 1]
```diff
@@ -120,6 +124,7 @@ public static void LogWrite(string message)
         // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
         public static void Configure(IApplicationBuilder app, IWebHostEnvironment env)
         {
+            Syncfusion.Licensing.SyncfusionLicenseProvider.RegisterLicense("NDQ0NzA0QDMxMzkyZTMxMmUzMGdaWENEdFBGbFZWYy9KWGF3aFZ6NmhFcnVZNkFvbHZnQ1pBNGVkMjAxNTg9");
```
Comment (Xian55) [inline BlazorServer/Startup.cs:127]: In general its recommended to include secret keys in the source code 😞

File: Core/Session/GrindingSession.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+        // this will change for TBC
+        // we might need to consider getting Wow client version and determine the calculation from there
+        private readonly double[] _experience =
+        {
+            400, 900, 1400, 2100, 2800, 3600, 4500, 5400, 6500, 7600, 8800, 10100, 11400, 12900, 14400, 16000, 17700,
+            19400, 21300, 23200, 25200, 27300, 29400, 31700, 34000, 36400, 38900, 41400, 44300, 47400, 50800, 54700,
+            58600, 62800, 67000, 71600, 76100, 80800, 85700, 90700, 95800, 101000, 106300, 111800, 117400, 123200,
+            129100, 135100, 141200, 147500, 153900, 160400, 167100, 173900, 180800, 187900, 195000, 202300, 209800
+        };
```
Comment (Xian55) [inline Core/Session/GrindingSession.cs:21]: This should be extracted to a config file under the `Json/xyz`

File: BlazorServer/Pages/_Host.cshtml [hunk 3]
```diff
@@ -8,28 +8,29 @@
...
+  <link href="css/site.css" rel="stylesheet" />
+  <script src="https://d3js.org/d3.v4.min.js"></script>
+  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.0.0/animate.min.css" />
+
+  <script>const whTooltips = { colorLinks: true, iconizeLinks: true, renameLinks: true };</script>
+  <script src="https://wow.zamimg.com/widgets/power.js"></script>
+
+  <script src="http://ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js"></script>
+  <script src="https://www.youtube.com/iframe_api"></script>
+  <script src="~/script/script.js"></script>
```
Comment (Xian55) [inline BlazorServer/Pages/_Host.cshtml:33]: if I'm not mistaken, only the indentation changed and `_content/Syncfusion.Blazor/styles/bootstrap-dark.css` added

File: BlazorServer/Startup.cs (hunk 1 above)
Comment (Xian55) [inline BlazorServer/Startup.cs:127]: This looking really sketchy.

In general its recommended to include secret keys in the source code :disappointed:

File: Core/BotController.cs [hunk 4]
```diff
@@ -7,9 +7,11 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using System.Net.Http.Headers;
```
Comment (Xian55) [inline Core/BotController.cs:10]: Which part of the code requires this namespace ?

File: Core/Session/GrindingSession.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        public double XpTo { get; set; }
+        public int MobsKilled { get; set; }
+        [JsonIgnore]
+        public double ExperiencePerHour => TotalTimeInMinutes == 0 ? 0 : Math.Round((double)(ExpGetInBotSession / TotalTimeInMinutes * 60), 0);
+        [JsonIgnore]
+        public double ExpGetInBotSession
+        {
+            get
+            {
+                if (LevelFrom == 60)
```
Comment (Xian55) [inline Core/Session/GrindingSession.cs:48]: This magic number `60` or `MaxLevel` should be based on the config file imho. Like counting all the mentioned experience to level number which immediately implies what the `MaxLevel` is

File: Core/Session/GrindingSession.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                    return 0;
+
+                if (LevelFrom == 59 && LevelTo == 60)
```
Comment (Xian55) [inline Core/Session/GrindingSession.cs:51]: similar to the past comment, `MaxLevel`

File: Core/Session/GrindingSession.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace Core.Session
+{
+    public class GrindingSession : IGrindingSession
+    {
+        private readonly IBotController _botController;
```
Comment (Xian55) [inline Core/Session/GrindingSession.cs:9]: Personally I don't like the naming, using the `_` as a prefix. in the constructor you can use the `this.` keyword instead.

File: Core/Session/IGrindingSession.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Core.Session
+{
+    public interface IGrindingSession
```
Comment (Xian55) [inline Core/Session/IGrindingSession.cs:6]: Can we rename everywhere `IGrindingSession` -> `IGrindSession`. `Grinding` sounds a little bit weird for me

File: Core/Session/LocalGrindingBotSessionHandler.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            // first time load
+            if (!Directory.Exists(_historyPath))
+            {
+                Directory.CreateDirectory(_historyPath);
+                return new List<GrindingSession>();
+            }
+                
+            var previousSessions =
+                Directory.EnumerateFiles($"{_historyPath}", "*.json")
```
Comment (Xian55) [inline Core/Session/LocalGrindingBotSessionHandler.cs:31]: Can you elaborate about this `$"{_historyPath}"` ? 🤔

File: Core/Session/LocalGrindingBotSessionHandler.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+            return previousSessions;
+        }
+
+        public void Save(IGrindingSession grindingSession)
+        {
+            var json = JsonConvert.SerializeObject(grindingSession);
+            if (!Directory.Exists(_historyPath))
+                Directory.CreateDirectory(_historyPath);
+            File.WriteAllText($@"{_historyPath}{grindingSession.SessionId}.json", json);
```
Comment (Xian55) [inline Core/Session/LocalGrindingBotSessionHandler.cs:44]: That's even more weird than the previous one 🤔 🤔

**Review by Xian55 — CHANGES_REQUESTED** [review]
In overall the PR looks really fleshed out, and nicely used the framework. 👏 

Just a few minor nitpicking, but nothing major core issue have been found.

Good work!!

File: BlazorServer/Pages/_Host.cshtml (hunk 3 above)
Comment (LarryLan) [reply BlazorServer/Pages/_Host.cshtml:33]: yes you're right mate

File: Core/Session/GrindingSession.cs (hunk 2 above)
Comment (LarryLan) [reply Core/Session/GrindingSession.cs:21]: good idea will be done.

File: Core/Session/LocalGrindingBotSessionHandler.cs (hunk 9 above)
Comment (LarryLan) [reply Core/Session/LocalGrindingBotSessionHandler.cs:31]: currently each grindsession is saved down in Json/History folder

**Comment by LarryLan** [comment]
> In overall the PR looks really fleshed out, and nicely used the framework. 👏
> 
> Just a few minor nitpicking, but nothing major core issue have been found.
> 
> Good work!!

thanks for the review mate.
I'll make the change and fix the bug tonight and send another one.

File: Core/Session/LocalGrindingBotSessionHandler.cs (hunk 9 above)
Comment (Xian55) [reply Core/Session/LocalGrindingBotSessionHandler.cs:31]: I mean the syntax what being used :)

File: Core/Session/GrindingSession.cs (hunk 7 above)
Comment (LarryLan) [reply Core/Session/GrindingSession.cs:9]: i'm getting used to underscore naming conventions for private method etc.
do you have fx cop or something like that i can change my coding style to fit to the repo.

Comment (Xian55) [reply Core/Session/GrindingSession.cs:9]: well there's no coding convention for the project which set in stone yet. Personally I've just tried to use the already set coding style.

File: Core/Session/LocalGrindingBotSessionHandler.cs (hunk 10 above)
Comment (Xian55) [reply Core/Session/LocalGrindingBotSessionHandler.cs:44]: the syntax look weird here as well
`$@"{_historyPath}{grindingSession.SessionId}.json"`

File: DataConfig/DataConfig.cs [hunk 11]
```diff
@@ -28,7 +28,8 @@ public class DataConfig
     public string Area => System.IO.Path.Join(Root, "area/");
     [JsonIgnore]
     public string PPather => System.IO.Path.Join(Root, "PPather/");
-
+    [JsonIgnore]
+    public string History => System.IO.Path.Join(Root, "History/");
```
Comment (Xian55) [inline DataConfig/DataConfig.cs:32]: Please increase the `DataConfigVersion.Version` value. Thanks!

File: PathingAPI/data_config.json [hunk 12]
```diff
@@ -0,0 +1 @@
+{"Version":7,"Root":"../json/"}
```
Comment (Xian55) [inline PathingAPI/data_config.json:1]: can you put the `data_config.json` file into the `.gitignore` list and remove this. please

File: Core/BotController.cs [hunk 13]
```diff
@@ -192,6 +200,9 @@ public void ToggleBotStatus()
                 else
                 {
                     actionThread.Active = false;
+                    this.GrindSession.StopBotSession();
```
Comment (Xian55) [inline Core/BotController.cs:203]: What happens to a `GrindSession` when the application does not exits gracefully? and this method and this
```
this.GrindSession.StopBotSession();
```
does not called?

File: PathingAPI/data_config.json (hunk 12 above)
Comment (LarryLan) [reply PathingAPI/data_config.json:1]: done

File: Core/BotController.cs (hunk 13 above)
Comment (LarryLan) [reply Core/BotController.cs:203]: it will record nothing for that session.

Comment (Xian55) [reply Core/BotController.cs:203]: okay just good to know

**Review by Xian55 — APPROVED** [review]

**Pull request merged**